File: src/monitor/_cpp/src/plugin_base.cpp
// FILE: src/monitor/_cpp/src/plugin_base.cpp
// VERSION: 1.0.0
// START_MODULE_CONTRACT:
// PURPOSE: Реализация базового класса плагина мониторинга на C++.
// SCOPE: Реализация методов PluginMetadata, MonitorAppProtocol, BaseMonitorPlugin
// INPUT: Заголовочный файл plugin_base.hpp
// OUTPUT: Скомпилированный объектный код
// KEYWORDS: [DOMAIN(9): Monitoring; DOMAIN(8): PluginSystem; TECH(6): Implementation]
// END_MODULE_CONTRACT

#include "plugin_base.hpp"
#include <pybind11/pybind11.h>
#include <pybind11/stl.h>
#include <pybind11/functional.h>
#include <ctime>
#include <iomanip>

namespace py = pybind11;

// START_CONSTANTS_DEFINITION
// Статические константы класса BaseMonitorPlugin
const std::string BaseMonitorPlugin::VERSION = "1.0.0";
const std::string BaseMonitorPlugin::AUTHOR = "Wallet Generator Team";
const std::string BaseMonitorPlugin::DESCRIPTION = "Base monitor plugin";
// END_CONSTANTS_DEFINITION


// START_PLUGIN_METADATA_IMPLEMENTATION

// START_CONSTRUCTOR_PLUGIN_METADATA_DEFAULT
// START_CONTRACT:
// PURPOSE: Конструктор по умолчанию для PluginMetadata
// OUTPUTS: Объект с пустыми значениями
// KEYWORDS: [CONCEPT(5): Initialization]
// END_CONTRACT
PluginMetadata::PluginMetadata()
    : name("")
    , version("")
    , description("")
    , author("")
    , dependencies({})
    , tags({})
    , homepage("")
    , license("")
{
}
// END_CONSTRUCTOR_PLUGIN_METADATA_DEFAULT


// START_CONSTRUCTOR_PLUGIN_METADATA_PARAMS
// START_CONTRACT:
// PURPOSE: Конструктор с параметрами для PluginMetadata
// INPUTS:
// - name: const std::string& — уникальное имя плагина
// - version: const std::string& — версия плагина
// - description: const std::string& — описание плагина
// - author: const std::string& — автор плагина
// - dependencies: const std::vector<std::string>& — зависимости плагина
// - tags: const std::vector<std::string>& — теги плагина
// - homepage: const std::string& — домашняя страница
// - license: const std::string& — лицензия
// OUTPUTS: Инициализированный объект PluginMetadata
// KEYWORDS: [CONCEPT(5): Initialization]
// END_CONTRACT
PluginMetadata::PluginMetadata(
    const std::string& name,
    const std::string& version,
    const std::string& description,
    const std::string& author,
    const std::vector<std::string>& dependencies,
    const std::vector<std::string>& tags,
    const std::string& homepage,
    const std::string& license)
    : name(name)
    , version(version)
    , description(description)
    , author(author)
    , dependencies(dependencies)
    , tags(tags)
    , homepage(homepage)
    , license(license)
{
}
// END_CONSTRUCTOR_PLUGIN_METADATA_PARAMS


// START_METHOD_PLUGIN_METADATA_FROM_PYTHON
// START_CONTRACT:
// PURPOSE: Создание PluginMetadata из Python объекта
// INPUTS:
// - obj: py::object — Python объект (dict или dataclass)
// OUTPUTS: PluginMetadata — созданные метаданные
// KEYWORDS: [PATTERN(7): Factory; CONCEPT(6): PythonBinding]
// END_CONTRACT
PluginMetadata PluginMetadata::from_python(py::object obj) {
    // Извлечение атрибутов из Python объекта
    std::string name = "";
    std::string version = "";
    std::string description = "";
    std::string author = "";
    std::vector<std::string> dependencies;
    std::vector<std::string> tags;
    std::string homepage = "";
    std::string license = "";

    // Попытка извлечения атрибутов
    try {
        if (py::hasattr(obj, "name")) {
            name = obj.attr("name").cast<std::string>();
        }
        if (py::hasattr(obj, "version")) {
            version = obj.attr("version").cast<std::string>();
        }
        if (py::hasattr(obj, "description")) {
            description = obj.attr("description").cast<std::string>();
        }
        if (py::hasattr(obj, "author")) {
            author = obj.attr("author").cast<std::string>();
        }
        if (py::hasattr(obj, "dependencies")) {
            dependencies = obj.attr("dependencies").cast<std::vector<std::string>>();
        }
        if (py::hasattr(obj, "tags")) {
            tags = obj.attr("tags").cast<std::vector<std::string>>();
        }
        if (py::hasattr(obj, "homepage")) {
            homepage = obj.attr("homepage").cast<std::string>();
        }
        if (py::hasattr(obj, "license")) {
            license = obj.attr("license").cast<std::string>();
        }
    } catch (const py::error_already_set& e) {
        // Логирование ошибки при извлечении атрибутов
        std::cerr << "[PluginMetadata][FROM_PYTHON][ExceptionCaught] Ошибка при извлечении атрибутов: " << e.what() << std::endl;
    }

    return PluginMetadata(name, version, description, author, dependencies, tags, homepage, license);
}
// END_METHOD_PLUGIN_METADATA_FROM_PYTHON

// END_PLUGIN_METADATA_IMPLEMENTATION


// START_MONITOR_APP_PROTOCOL_IMPLEMENTATION


// START_CONSTRUCTOR_MONITOR_APP_PROTOCOL
// START_CONTRACT:
// PURPOSE: Конструктор MonitorAppProtocol
// INPUTS:
// - app: py::object — ссылка на главное приложение мониторинга
// OUTPUTS: Инициализированный объект протокола
// KEYWORDS: [CONCEPT(5): Initialization]
// END_CONTRACT
MonitorAppProtocol::MonitorAppProtocol(py::object app)
    : _app(app)
{
}
// END_CONSTRUCTOR_MONITOR_APP_PROTOCOL


// START_METHOD_REGISTER_UI_COMPONENT
// START_CONTRACT:
// PURPOSE: Регистрация UI компонента в указанной вкладке
// INPUTS:
// - component: py::object — UI компонент для регистрации
// - tab_name: const std::string& — имя вкладки для размещения компонента
// OUTPUTS: void
// SIDE_EFFECTS: Вызывает метод register_ui_component у приложения
// KEYWORDS: [DOMAIN(7): UI; CONCEPT(6): Registration]
// END_CONTRACT
void MonitorAppProtocol::register_ui_component(py::object component, const std::string& tab_name) {
    if (py::hasattr(_app, "register_ui_component")) {
        try {
            _app.attr("register_ui_component")(component, tab_name);
        } catch (const py::error_already_set& e) {
            std::cerr << "[MonitorAppProtocol][REGISTER_UI_COMPONENT][ExceptionCaught] Ошибка при регистрации UI: " << e.what() << std::endl;
        }
    }
}
// END_METHOD_REGISTER_UI_COMPONENT


// START_METHOD_LOG
// START_CONTRACT:
// PURPOSE: Логирование сообщения с указанным уровнем
// INPUTS:
// - level: const std::string& — уровень логирования
// - message: const std::string& — текст сообщения
// OUTPUTS: void
// KEYWORDS: [CONCEPT(5): Logging]
// END_CONTRACT
void MonitorAppProtocol::log(const std::string& level, const std::string& message) {
    if (py::hasattr(_app, "log")) {
        try {
            _app.attr("log")(level, message);
        } catch (const py::error_already_set& e) {
            std::cerr << "[MonitorAppProtocol][LOG][ExceptionCaught] Ошибка при логировании: " << e.what() << std::endl;
        }
    }
}
// END_METHOD_LOG


// START_METHOD_GET_METRICS_STORE
// START_CONTRACT:
// PURPOSE: Получение хранилища метрик от приложения
// OUTPUTS: py::object — хранилище метрик или None
// KEYWORDS: [DOMAIN(8): MetricsStore; CONCEPT(6): Getter]
// END_CONTRACT
py::object MonitorAppProtocol::get_metrics_store() {
    if (py::hasattr(_app, "get_metrics_store")) {
        try {
            return _app.attr("get_metrics_store")();
        } catch (const py::error_already_set& e) {
            std::cerr << "[MonitorAppProtocol][GET_METRICS_STORE][ExceptionCaught] Ошибка при получении метрик: " << e.what() << std::endl;
        }
    }
    return py::none();
}
// END_METHOD_GET_METRICS_STORE

// END_MONITOR_APP_PROTOCOL_IMPLEMENTATION


// START_BASE_MONITOR_PLUGIN_IMPLEMENTATION


// START_CONSTRUCTOR_BASE_MONITOR_PLUGIN
// START_CONTRACT:
// PURPOSE: Конструктор базового плагина с именем и приоритетом
// INPUTS:
// - name: const std::string& — уникальное имя плагина
// - priority: int — приоритет плагина
// OUTPUTS: Инициализированный объект плагина
// SIDE_EFFECTS: Создает внутреннее состояние плагина; устанавливает начальные значения атрибутов
// KEYWORDS: [CONCEPT(5): Initialization; DOMAIN(7): PluginSetup]
// END_CONTRACT
BaseMonitorPlugin::BaseMonitorPlugin(const std::string& name, int priority)
    : _name(name)
    , _priority(priority)
    , _enabled(true)
    , _app(py::none())
{
    // Логирование инициализации
    std::stringstream ss;
    ss << "[BaseMonitorPlugin][INIT][ConditionCheck] Инициализирован плагин: " << name << " с приоритетом " << priority;
    log_debug(ss.str());
}
// END_CONSTRUCTOR_BASE_MONITOR_PLUGIN


// START_METHOD_GET_UI_COMPONENTS
// START_CONTRACT:
// PURPOSE: Получение UI компонентов плагина для отображения (базовая реализация)
// OUTPUTS: py::object — None по умолчанию
// KEYWORDS: [PATTERN(6): UI; CONCEPT(5): Getter]
// END_CONTRACT
py::object BaseMonitorPlugin::get_ui_components() {
    return py::none();
}
// END_METHOD_GET_UI_COMPONENTS


// START_METHOD_GET_METADATA
// START_CONTRACT:
// PURPOSE: Получение метаданных плагина
// OUTPUTS: PluginMetadata — метаданные плагина
// KEYWORDS: [CONCEPT(5): Getter; DOMAIN(8): PluginMetadata]
// END_CONTRACT
PluginMetadata BaseMonitorPlugin::get_metadata() const {
    return PluginMetadata(
        _name,
        VERSION,
        DESCRIPTION,
        AUTHOR
    );
}
// END_METHOD_GET_METADATA


// START_METHOD_ON_START
// START_CONTRACT:
// PURPOSE: Обработчик запуска генератора кошельков (базовая реализация)
// INPUTS:
// - selected_list_path: const std::string& — путь к выбранному списку адресов
// OUTPUTS: void
// SIDE_EFFECTS: Логирует событие запуска
// KEYWORDS: [DOMAIN(8): EventHandler; CONCEPT(6): Startup]
// END_CONTRACT
void BaseMonitorPlugin::on_start(const std::string& selected_list_path) {
    std::stringstream ss;
    ss << "[BaseMonitorPlugin][ON_START][StepComplete] Плагин " << _name << " получил уведомление о запуске";
    log_debug(ss.str());
}
// END_METHOD_ON_START


// START_METHOD_ON_MATCH_FOUND
// START_CONTRACT:
// PURPOSE: Обработчик обнаружения совпадений (базовая реализация)
// INPUTS:
// - matches: py::list — список найденных совпадений
// - iteration: int — номер итерации
// OUTPUTS: void
// KEYWORDS: [DOMAIN(9): MatchHandler; CONCEPT(7): EventHandler]
// END_CONTRACT
void BaseMonitorPlugin::on_match_found(py::list matches, int iteration) {
    std::stringstream ss;
    ss << "[BaseMonitorPlugin][ON_MATCH_FOUND][Info] Найдено совпадений: " << py::len(matches) << " на итерации " << iteration;
    log_info(ss.str());
}
// END_METHOD_ON_MATCH_FOUND


// START_METHOD_ON_FINISH
// START_CONTRACT:
// PURPOSE: Обработчик завершения генерации кошельков (базовая реализация)
// INPUTS:
// - final_metrics: py::dict — финальные метрики генерации
// OUTPUTS: void
// KEYWORDS: [DOMAIN(8): Finalization; CONCEPT(7): EventHandler]
// END_CONTRACT
void BaseMonitorPlugin::on_finish(py::dict final_metrics) {
    std::stringstream ss;
    ss << "[BaseMonitorPlugin][ON_FINISH][StepComplete] Плагин " << _name << " получил финальные метрики";
    log_debug(ss.str());
}
// END_METHOD_ON_FINISH


// START_METHOD_ON_RESET
// START_CONTRACT:
// PURPOSE: Обработчик сброса генератора (базовая реализация)
// OUTPUTS: void
// KEYWORDS: [DOMAIN(7): ResetHandler; CONCEPT(7): EventHandler]
// END_CONTRACT
void BaseMonitorPlugin::on_reset() {
    std::stringstream ss;
    ss << "[BaseMonitorPlugin][ON_RESET][StepComplete] Плагин " << _name << " получил уведомление о сбросе";
    log_debug(ss.str());
}
// END_METHOD_ON_RESET


// START_METHOD_ENABLE
// START_CONTRACT:
// PURPOSE: Включение плагина
// OUTPUTS: void
// SIDE_EFFECTS: Устанавливает флаг enabled = true; логирует событие включения
// KEYWORDS: [CONCEPT(6): StateManagement]
// END_CONTRACT
void BaseMonitorPlugin::enable() {
    _enabled = true;
    std::stringstream ss;
    ss << "[BaseMonitorPlugin][ENABLE][StepComplete] Плагин " << _name << " включён";
    log_info(ss.str());
}
// END_METHOD_ENABLE


// START_METHOD_DISABLE
// START_CONTRACT:
// PURPOSE: Отключение плагина
// OUTPUTS: void
// SIDE_EFFECTS: Устанавливает флаг enabled = false; логирует событие отключения
// KEYWORDS: [CONCEPT(6): StateManagement]
// END_CONTRACT
void BaseMonitorPlugin::disable() {
    _enabled = false;
    std::stringstream ss;
    ss << "[BaseMonitorPlugin][DISABLE][StepComplete] Плагин " << _name << " отключён";
    log_info(ss.str());
}
// END_METHOD_DISABLE


// START_METHOD_IS_ENABLED
// START_CONTRACT:
// PURPOSE: Проверка состояния плагина
// OUTPUTS: bool — true если плагин включён, false если выключен
// KEYWORDS: [CONCEPT(5): Getter; DOMAIN(6): StateCheck]
// END_CONTRACT
bool BaseMonitorPlugin::is_enabled() const {
    return _enabled;
}
// END_METHOD_IS_ENABLED


// START_METHOD_HEALTH_CHECK
// START_CONTRACT:
// PURPOSE: Проверка работоспособности плагина (базовая реализация)
// OUTPUTS: bool — true (всегда работает корректно по умолчанию)
// KEYWORDS: [CONCEPT(7): HealthCheck; DOMAIN(6): Diagnostics]
// END_CONTRACT
bool BaseMonitorPlugin::health_check() const {
    return true;
}
// END_METHOD_HEALTH_CHECK


// START_METHOD_GET_NAME
// START_CONTRACT:
// PURPOSE: Получение имени плагина
// OUTPUTS: std::string — имя плагина
// KEYWORDS: [CONCEPT(5): Getter]
// END_CONTRACT
std::string BaseMonitorPlugin::get_name() const {
    return _name;
}
// END_METHOD_GET_NAME


// START_METHOD_GET_PRIORITY
// START_CONTRACT:
// PURPOSE: Получение приоритета плагина
// OUTPUTS: int — приоритет плагина
// KEYWORDS: [CONCEPT(5): Getter]
// END_CONTRACT
int BaseMonitorPlugin::get_priority() const {
    return _priority;
}
// END_METHOD_GET_PRIORITY


// START_METHOD_SET_PRIORITY
// START_CONTRACT:
// PURPOSE: Установка приоритета
// INPUTS:
// - priority: int — новый приоритет
// OUTPUTS: void
// KEYWORDS: [CONCEPT(6): Setter]
// END_CONTRACT
void BaseMonitorPlugin::set_priority(int priority) {
    _priority = priority;
    std::stringstream ss;
    ss << "[BaseMonitorPlugin][SET_PRIORITY][StepComplete] Приоритет плагина " << _name << " изменён на " << priority;
    log_debug(ss.str());
}
// END_METHOD_SET_PRIORITY


// START_METHOD_GET_ENABLED
// START_CONTRACT:
// PURPOSE: Получение состояния enabled (геттер)
// OUTPUTS: bool — состояние включения плагина
// KEYWORDS: [CONCEPT(5): Getter]
// END_CONTRACT
bool BaseMonitorPlugin::get_enabled() const {
    return _enabled;
}
// END_METHOD_GET_ENABLED


// START_METHOD_SET_ENABLED
// START_CONTRACT:
// PURPOSE: Установка состояния enabled (сеттер)
// INPUTS:
// - enabled: bool — новое состояние включения
// OUTPUTS: void
// KEYWORDS: [CONCEPT(6): Setter]
// END_CONTRACT
void BaseMonitorPlugin::set_enabled(bool enabled) {
    _enabled = enabled;
}
// END_METHOD_SET_ENABLED


// START_OPERATOR_LT_IMPLEMENTATION
// START_CONTRACT:
// PURPOSE: Сравнение плагинов по приоритету (оператор <)
// INPUTS:
// - other: const BaseMonitorPlugin& — другой плагин для сравнения
// OUTPUTS: bool — результат сравнения
// KEYWORDS: [CONCEPT(8): Comparison; DOMAIN(6): Sorting]
// END_CONTRACT
bool BaseMonitorPlugin::operator<(const BaseMonitorPlugin& other) const {
    return _priority < other._priority;
}
// END_OPERATOR_LT_IMPLEMENTATION


// START_OPERATOR_EQ_IMPLEMENTATION
// START_CONTRACT:
// PURPOSE: Проверка равенства плагинов по имени
// INPUTS:
// - other: const BaseMonitorPlugin& — объект для сравнения
// OUTPUTS: bool — результат сравнения
// KEYWORDS: [CONCEPT(7): Equality; DOMAIN(5): Comparison]
// END_CONTRACT
bool BaseMonitorPlugin::operator==(const BaseMonitorPlugin& other) const {
    return _name == other._name;
}
// END_OPERATOR_EQ_IMPLEMENTATION


// START_METHOD_HASH_IMPLEMENTATION
// START_CONTRACT:
// PURPOSE: Получение хэша плагина на основе имени
// OUTPUTS: size_t — хэш плагина
// KEYWORDS: [CONCEPT(6): Hashing]
// END_CONTRACT
size_t BaseMonitorPlugin::hash() const {
    return std::hash<std::string>()(_name);
}
// END_METHOD_HASH_IMPLEMENTATION


// START_LOGGING_METHODS

// START_METHOD_LOG_DEBUG
// START_CONTRACT:
// PURPOSE: Внутренний метод для логирования отладочных сообщений
// INPUTS:
// - message: const std::string& — текст сообщения
// OUTPUTS: void
// KEYWORDS: [CONCEPT(5): Logging]
// END_CONTRACT
void BaseMonitorPlugin::log_debug(const std::string& message) {
    std::cout << get_timestamp() << " [DEBUG] " << message << std::endl;
}
// END_METHOD_LOG_DEBUG


// START_METHOD_LOG_INFO
// START_CONTRACT:
// PURPOSE: Внутренний метод для логирования информационных сообщений
// INPUTS:
// - message: const std::string& — текст сообщения
// OUTPUTS: void
// KEYWORDS: [CONCEPT(5): Logging]
// END_CONTRACT
void BaseMonitorPlugin::log_info(const std::string& message) {
    std::cout << get_timestamp() << " [INFO] " << message << std::endl;
}
// END_METHOD_LOG_INFO


// START_METHOD_LOG_WARNING
// START_CONTRACT:
// PURPOSE: Внутренний метод для логирования предупреждений
// INPUTS:
// - message: const std::string& — текст сообщения
// OUTPUTS: void
// KEYWORDS: [CONCEPT(5): Logging]
// END_CONTRACT
void BaseMonitorPlugin::log_warning(const std::string& message) {
    std::cout << get_timestamp() << " [WARNING] " << message << std::endl;
}
// END_METHOD_LOG_WARNING


// START_METHOD_LOG_ERROR
// START_CONTRACT:
// PURPOSE: Внутренний метод для логирования ошибок
// INPUTS:
// - message: const std::string& — текст сообщения
// OUTPUTS: void
// KEYWORDS: [CONCEPT(5): Logging]
// END_CONTRACT
void BaseMonitorPlugin::log_error(const std::string& message) {
    std::cout << get_timestamp() << " [ERROR] " << message << std::endl;
}
// END_METHOD_LOG_ERROR


// START_HELPER_GET_TIMESTAMP
// START_CONTRACT:
// PURPOSE: Получение текущей временной метки в формате для логирования
// OUTPUTS: std::string — отформатированная временная метка
// KEYWORDS: [CONCEPT(6): Timestamp]
// END_CONTRACT
std::string BaseMonitorPlugin::get_timestamp() {
    auto now = std::time(nullptr);
    auto tm = *std::localtime(&now);
    std::ostringstream oss;
    oss << std::put_time(&tm, "%Y-%m-%d %H:%M:%S");
    return oss.str();
}
// END_HELPER_GET_TIMESTAMP

// END_BASE_MONITOR_PLUGIN_IMPLEMENTATION

File: src/monitor/_cpp/plugin_base.py
# FILE: src/monitor/_cpp/plugin_base.py
# VERSION: 2.0.0
# START_MODULE_CONTRACT:
# PURPOSE: Python-обёртка для C++ модуля базового класса плагинов мониторинга. Обеспечивает единый интерфейс для создания плагинов. Только C++ реализация - Python fallback НЕПРИЕМЛЕМ.
# SCOPE: Абстрактный базовый класс, система метаданных плагинов, протокол приложения мониторинга
# INPUT: Нет (модуль предоставляет классы)
# OUTPUT: Классы BaseMonitorPlugin, PluginMetadata, MonitorAppProtocol
# KEYWORDS: [DOMAIN(9): PluginSystem; DOMAIN(8): CppBinding; CONCEPT(9): NoFallback; TECH(7): PyBind11]
# LINKS: [IMPORTS(9): plugin_base_cpp; EXPORTS(9): BaseMonitorPlugin; EXPORTS(8): PluginMetadata; EXPORTS(8): MonitorAppProtocol]
# LINKS_TO_SPECIFICATION: [Спецификация плагинов мониторинга из docs/cpp_modules/plugin_base_draft_code_plan_v2.md]
# END_MODULE_CONTRACT
# START_MODULE_MAP:
# CLASS 10 [Базовый класс для всех плагинов мониторинга] => BaseMonitorPlugin
# CLASS 8 [Метаданные плагина] => PluginMetadata
# CLASS 7 [Протокол приложения мониторинга] => MonitorAppProtocol
# END_MODULE_MAP
# START_USE_CASES:
# - [BaseMonitorPlugin]: System (PluginLifecycle) -> ProvideBaseInterface -> PluginArchitectureDefined
# - [PluginMetadata]: Plugin (Registration) -> ProvideMetadata -> PluginInfoAvailable
# - [MonitorAppProtocol]: App (Communication) -> DefineProtocol -> InterProcessCommunicationReady
# END_USE_CASES
"""
Модуль plugin_base.py — Python-обёртка для C++ модуля базового класса плагинов мониторинга.

ВНИМАНИЕ: Этот модуль требует наличия скомпилированного C++ модуля plugin_base_cpp.
Python fallback НЕПРЕДУСМОТРЕН - только C++ реализация.

Экспортирует:
    - BaseMonitorPlugin: Базовый класс для всех плагинов мониторинга
    - PluginMetadata: Класс для хранения метаданных плагина
    - MonitorAppProtocol: Класс протокола приложения мониторинга

При отсутствии C++ модуля выбрасывается ImportError с инструкцией по сборке.
"""

import logging
from pathlib import Path

# ============================================================================
# КОНФИГУРАЦИЯ ЛОГИРОВАНИЯ
# ============================================================================

def _setup_file_logging() -> logging.FileHandler:
    """Настройка логирования в файл app.log."""
    project_root = Path(__file__).parent.parent.parent
    log_file = project_root / "app.log"
    file_handler = logging.FileHandler(log_file, mode='a', encoding='utf-8')
    file_handler.setLevel(logging.INFO)
    file_handler.setFormatter(logging.Formatter(
        '[%(asctime)s][%(name)s][%(levelname)s] %(message)s',
        datefmt='%Y-%m-%d %H:%M:%S'
    ))
    return file_handler

# Настройка логирования для модуля
_logger = logging.getLogger("src.monitor._cpp.plugin_base")
if not _logger.handlers:
    _file_handler = _setup_file_logging()
    _logger.addHandler(_file_handler)
    _logger.setLevel(logging.INFO)

# ============================================================================
# ИМПОРТ C++ МОДУЛЯ
# ============================================================================

# START_FUNCTION_IMPORT_CPP_MODULE
# START_CONTRACT:
# PURPOSE: Импорт C++ модуля plugin_base_cpp с жёсткой проверкой доступности
# INPUTS: Нет
# OUTPUTS: Импортированный модуль
# SIDE_EFFECTS: Вызывает ImportError при отсутствии C++ модуля
# KEYWORDS: [PATTERN(9): HardImport; DOMAIN(9): ModuleLoading; TECH(6): Import]
# END_CONTRACT

def _import_cpp_module():
    """
    Импорт C++ модуля plugin_base_cpp.
    
    ВНИМАНИЕ: Это жёсткий импорт. При отсутствии C++ модуля
    выбрасывается ImportError с инструкцией по сборке.
    
    Returns:
        Импортированный C++ модуль plugin_base_cpp
        
    Raises:
        ImportError: Если C++ модуль недоступен
    """
    _logger.info(f"[IMPORT][START] Попытка импорта plugin_base_cpp...")
    
    try:
        import plugin_base_cpp
        _logger.info(f"[IMPORT][SUCCESS] C++ модуль plugin_base_cpp успешно импортирован [SUCCESS]")
        return plugin_base_cpp
    except (ImportError, ModuleNotFoundError) as e:
        _logger.error(f"[IMPORT][CRITICAL] C++ модуль plugin_base_cpp недоступен: {e} [FAIL]")
        
        build_instructions = f"""
================================================================================
ОШИБКА: C++ модуль plugin_base_cpp недоступен
================================================================================

Требуемый модуль: plugin_base_cpp
Описание: Базовый класс плагинов мониторинга

ИНСТРУКЦИЯ ПО СБОРКЕ:
----------------------
1. Перейдите в директорию src/monitor/_cpp
2. Выполните сборку C++ модулей:
   cd src/monitor/_cpp
   ./build.sh
   
   Или используя CMake:
   cd src/monitor/_cpp
   mkdir -p build && cd build
   cmake ..
   make

3. После успешной сборки перезапустите приложение

Документация: docs/cpp_modules/plugin_base_draft_code_plan_v2.md

================================================================================
"""
        _logger.error(f"[IMPORT][CRITICAL]{build_instructions}")
        raise ImportError(
            f"C++ модуль plugin_base_cpp недоступен. "
            f"Пожалуйста, соберите C++ модуль следуя инструкциям в docs/cpp_modules/plugin_base_draft_code_plan_v2.md"
        )

# END_FUNCTION_IMPORT_CPP_MODULE

# ============================================================================
# ИМПОРТ КЛАССОВ ИЗ C++ МОДУЛЯ
# ============================================================================

# Получаем C++ модуль
_cpp_module = _import_cpp_module()

# Импортируем классы из C++ модуля
try:
    BaseMonitorPlugin = _cpp_module.BaseMonitorPlugin
    _logger.info(f"[IMPORT][SUCCESS] Класс BaseMonitorPlugin импортирован из C++ [SUCCESS]")
except AttributeError as e:
    _logger.error(f"[IMPORT][CRITICAL] Класс BaseMonitorPlugin не найден в C++ модуле: {e} [FAIL]")
    raise ImportError(f"C++ модуль plugin_base_cpp не содержит класса BaseMonitorPlugin: {e}")

try:
    PluginMetadata = _cpp_module.PluginMetadata
    _logger.info(f"[IMPORT][SUCCESS] Класс PluginMetadata импортирован из C++ [SUCCESS]")
except AttributeError as e:
    _logger.error(f"[IMPORT][CRITICAL] Класс PluginMetadata не найден в C++ модуле: {e} [FAIL]")
    raise ImportError(f"C++ модуль plugin_base_cpp не содержит класса PluginMetadata: {e}")

try:
    MonitorAppProtocol = _cpp_module.MonitorAppProtocol
    _logger.info(f"[IMPORT][SUCCESS] Класс MonitorAppProtocol импортирован из C++ [SUCCESS]")
except AttributeError as e:
    _logger.warning(f"[IMPORT][WARNING] Класс MonitorAppProtocol не найден в C++ модуле: {e} [WARNING]")
    # MonitorAppProtocol может быть опциональным
    MonitorAppProtocol = None

# ============================================================================
# ЭКСПОРТ
# ============================================================================

__all__ = [
    # Классы из C++ модуля
    "BaseMonitorPlugin",
    "PluginMetadata",
    "MonitorAppProtocol",
]

# ============================================================================
# ИНИЦИАЛИЗАЦИЯ ПРИ ИМПОРТЕ
# ============================================================================

_logger.info(f"[INIT][SUCCESS] Модуль src.monitor._cpp.plugin_base успешно инициализирован, используется C++ реализация [SUCCESS]")
